external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1792 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(not(feature = "gas-profiler"))]
+pub(crate) use vm_tracer_noop::*;
+mod vm_tracer_noop {
+    use move_trace_format::trace_format::MoveTraceBuilder;
+
+    #[allow(dead_code)]
+    pub(crate) struct VMTracer<'a> {
+        _phantom: std::marker::PhantomData<&'a ()>,
+    }
+
+    impl<'a> VMTracer<'a> {
+        #[allow(dead_code)]
+        pub(crate) fn new(_trace: &'a mut MoveTraceBuilder) -> Self {
+            Self {
+                _phantom: std::marker::PhantomData,
+            }
+        }
+    }
+}
+
+#[cfg(feature = "gas-profiler")]
+pub(crate) use vm_tracer_impl::*;
+#[cfg(feature = "gas-profiler")]
+mod vm_tracer_impl {
+    use crate::{
+        interpreter::{Frame, Interpreter},
+        loader::{Function, Loader},
+    };
+    use move_binary_format::{
+        errors::PartialVMError,
+        file_format::{ConstantPoolIndex, SignatureIndex},
+    };
+    use move_core_types::{
+        account_address::AccountAddress,
+        annotated_value::{MoveTypeLayout, MoveValue},
+        language_storage::TypeTag,
+    };
+    use move_trace_format::trace_format::{
+        DataLoad, Effect as IF, Location, MoveTraceBuilder, Read, RefType, TraceIndex, TraceValue,
+        TypeTagWithRefs, Write,
+    };
+    use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+    use std::collections::BTreeMap;
+
+    /// Internal state for the TypeTracer. This is where the actual tracing logic is implemented.
+    pub(crate) struct VMTracer<'a> {
+        trace: &'a mut MoveTraceBuilder,
```

Comment:
Is there a compelling reason to borrow this value, instead of having the tracer hold and then return it?

Reply:
Main reason is passing it by value would make the interface for this and the VM more cumbersome since this would now affect the return types of the VM as well as input types (and call sites etc), and this feels more natural to me/descriptive of "who owns the tracer."